custom_model_runner/datarobot_drum/resource/components/Python/generic_predictor/generic_predictor.py

```diff
@@ -105,7 +106,13 @@ def _materialize(self, parent_data_objs, user_data):
             binary_data, mimetype = StructuredInputReadUtils.read_structured_input_file_as_binary(
                 input_filename
             )
-            transformed_df = self._predictor.transform(binary_data=binary_data, mimetype=mimetype)
+            transformed_output = self._predictor.transform(
```

Comment:
can we make `transform` always return either dataframe or tuple?
what is the reason it's sometimes dataframe and sometimes tuple

Reply:
its dependent on sending `y` to the transform. I agree though, we should just always make it a tuple

Reply:
so i'm curious on how to move forward with this.  To make it always return a tuple would require changes to model_adapter.py and we could essentially undo everything that was done to generic_predictor.  

part of model_adapter to change 👇 
```
        if self._custom_hooks.get(CustomHooks.TRANSFORM):
            try:
                transform_hook = self._custom_hooks[CustomHooks.TRANSFORM]
                transform_params = signature(transform_hook).parameters
                if len(transform_params) == 3:
                    transform_out = transform_hook(data, model, target_data)
                elif len(transform_params) == 2:
                    transform_out = transform_hook(data, model)
                else:
                    raise ValueError(
                        "Transform hook must take 2 or 3 arguments; "
                        "hook provided takes {}".format(len(transform_params))
                    )
                if type(transform_out) == tuple:
                    output_data, output_target = transform_out
                else:
                    output_data = transform_out
                    output_target = target_data

```
@yakov-g thoughts??

Reply:
So, as I understand, this solution won't work as `transform` which is defined here,
https://github.com/datarobot/datarobot-user-models/blob/a91996cc2674162424924cffe6e4de97e75a03e6/custom_model_runner/datarobot_drum/drum/language_predictors/base_language_predictor.py#L123-L128
will always return tuple, returned all the way from model_adapter (even if the hook returns a single value). So we already always return tuple. 
Thus, looks like, all that has to be done is adding `[0]`: 
`transformed_df = self._predictor.transform(binary_data=binary_data, mimetype=mimetype)[0]`